toolchain/parse/handle_decl_scope_loop.cpp

```diff
@@ -88,118 +54,112 @@ static auto ApplyIntroducer(Context& context, Context::StateStackEntry state,
+  return std::to_array(introducers);
+}();
+}  // namespace
+
+// Attempts to handle the current token as a declaration introducer.
 // Returns true if the current position is a declaration. If we see a
 // declaration introducer keyword token, replace the placeholder node and switch
 // to a state to parse the rest of the declaration.
 static auto TryHandleAsDecl(Context& context, Context::StateStackEntry state,
                             bool saw_modifier) -> bool {
-  switch (context.PositionKind()) {
-    case Lex::TokenKind::Adapt: {
-      ApplyIntroducer(context, state, NodeKind::AdaptIntroducer,
-                      State::AdaptDecl);
-      context.PushState(State::Expr);
-      return true;
-    }
-    case Lex::TokenKind::Alias: {
-      ApplyIntroducer(context, state, NodeKind::AliasIntroducer, State::Alias);
-      return true;
-    }
-    case Lex::TokenKind::Base: {
-      HandleBaseAsDecl(context, state);
-      return true;
-    }
-    case Lex::TokenKind::Choice: {
-      ApplyIntroducer(context, state, NodeKind::ChoiceIntroducer,
-                      State::ChoiceIntroducer);
-      return true;
-    }
-    case Lex::TokenKind::Class: {
-      ApplyIntroducer(context, state, NodeKind::ClassIntroducer,
-                      State::TypeAfterIntroducerAsClass);
-      return true;
-    }
-    case Lex::TokenKind::Constraint: {
-      ApplyIntroducer(context, state, NodeKind::NamedConstraintIntroducer,
-                      State::TypeAfterIntroducerAsNamedConstraint);
-      return true;
-    }
-    case Lex::TokenKind::Export: {
-      ApplyIntroducer(context, state, NodeKind::ExportIntroducer,
-                      State::ExportName);
-      return true;
-    }
-    case Lex::TokenKind::Extend: {
-      // TODO: Treat this `extend` token as a declaration introducer
-      HandleUnrecognizedDecl(context, state.subtree_start);
-      return true;
-    }
-    case Lex::TokenKind::Fn: {
-      ApplyIntroducer(context, state, NodeKind::FunctionIntroducer,
-                      State::FunctionIntroducer);
-      return true;
-    }
-    case Lex::TokenKind::Impl: {
-      ApplyIntroducer(context, state, NodeKind::ImplIntroducer,
-                      State::ImplAfterIntroducer);
-      return true;
-    }
-    case Lex::TokenKind::Interface: {
-      ApplyIntroducer(context, state, NodeKind::InterfaceIntroducer,
-                      State::TypeAfterIntroducerAsInterface);
-      return true;
-    }
-    case Lex::TokenKind::Namespace: {
-      ApplyIntroducer(context, state, NodeKind::NamespaceStart,
-                      State::Namespace);
-      return true;
-    }
-    case Lex::TokenKind::Let: {
-      ApplyIntroducer(context, state, NodeKind::LetIntroducer, State::Let);
-      return true;
+  const auto& info = DeclIntroducers[context.PositionKind().AsInt()];
+
+  switch (info.introducer_kind) {
+    case DeclIntroducerKind::Unrecognized: {
+      // A `;` with no modifiers is an empty declaration.
```

Comment:
Maybe a minor consequence of this: do we test / want to support `;` on its own at the top of a file? I think it would've become a packaging state error before. However, now you handle this _after_ packaging keywords, and so I think it's becoming valid...?

Note, this probably isn't tested, but maybe it should be. e.g.:

```
;

package Foo;
```

Reply:
Great point! Added tests.

We still reject a `;` before `package` and `library` declarations, but we now allow a `;` before or between `import` declarations. I hadn't thought about this before your comment, but now I have, I think that's the behavior that I would want here.